DylibBinder/DBTypeDeclaration.cs

```diff
@@ -0,0 +1,200 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using SwiftReflector.Inventory;
+using SwiftReflector.SwiftXmlReflection;
+
+namespace DylibBinder {
+	internal class DBTypeDeclaration {
+		public string Name { get; }
+		public string Module { get; }
+		public bool IsObjC { get; } = false;
+		public bool IsFinal { get; } = false;
+		public bool IsDeprecated { get; } = false;
+		public bool IsUnavailable { get; } = false;
+		public TypeKind Kind { get; }
+		public TypeAccessibility Accessibility { get; } = TypeAccessibility.Public;
+
+		public DBFuncs Funcs { get; }
+		public DBProperties Properties { get; }
+		public DBGenericParameters GenericParameters { get; }
+		public DBAssociatedTypes AssociatedTypes { get; }
+		public DBInnerTypes InnerTypes { get; }
+
+		public DBTypeDeclaration (ProtocolContents protoContents)
+		{
+			Kind = TypeKind.Protocol;
+			Name = protoContents.Name.ToFullyQualifiedName ();
+			Module = protoContents.Name.Module.Name;
+			Funcs = new DBFuncs (protoContents);
+			AssociatedTypes = new DBAssociatedTypes (this);
+			ApplyInouts ();
+		}
+
+		public DBTypeDeclaration (ClassContents classContents)
+		{
+			Kind = classContents.Name.IsClass ? TypeKind.Class : classContents.Name.IsEnum ? TypeKind.Enum : TypeKind.Struct;
+			Name = classContents.Name.ToFullyQualifiedName (false);
+			Module = classContents.Name.Module.Name;
+			Funcs = new DBFuncs (classContents);
+			Properties = new DBProperties (classContents);
+			GenericParameters = new DBGenericParameters (this);
+			InnerTypes = new DBInnerTypes (classContents.Name);
+			AssociatedTypes = new DBAssociatedTypes (this);
+			ApplyInouts ();
+		}
+
+		// we want to go through all the parameters in this DBTypeDeclaration
+		// and add the appropriate type to Instances and Constructors
+		void ApplyInouts ()
+		{
+			foreach (var func in Funcs.FuncCollection) {
+				ApplyInouts (func.ParameterLists);
+			}
+			// protocols currently only hold Funcs so they should finish here
+			if (Kind == TypeKind.Protocol)
+				return;
+
+			foreach (var property in Properties.PropertyCollection) {
+				ApplyInouts (property.Getter.ParameterLists);
+				if (property.Setter != null) {
+					ApplyInouts (property.Setter.ParameterLists);
+				}
+			}
+		}
+
+		void ApplyInouts (DBParameterLists parameterLists)
+		{
+			foreach (var parameterList in parameterLists.ParameterListCollection) {
+				ApplyInouts (parameterList);
+			}
+		}
+
+		void ApplyInouts (DBParameterList parameterList)
+		{
+			foreach (var parameter in parameterList.ParameterCollection) {
+				if (parameter.HasInstance || parameter.IsConstructor) {
+					ApplyInouts (parameter);
+				}
+			}
+		}
+
+		void ApplyInouts (DBParameter parameter)
+		{
+			var sb = new StringBuilder ();
+			sb.Append ($"inout {Module}.{Name}{GenericParametersToString ()}");
```

Comment:
the nice thing about `StringBuilder` is to be able to reduce temporary string allocations, like what `String.Format` would do. However if you use `$` you end up (very likely) with a `String.Format` call

Instead do

```csharp
sb.Append ("inout ").Append (module).Append ('.').Append (Name).Append (GenericParametersToString ());
```